File: leetcode/spiral-matrix.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])

        def traverseRow(row, r, result):
            for i in r:
                result.append(matrix[row][i])

        def traverseColumn(column, r, result):
            for i in r:
                result.append(matrix[i][column])

        res = []
        r = 0
        c = 0

        reverse = False

        while r < m and c < n:
            if reverse:
                traverseRow(m - 1, range(n - 1, c - 1, -1), res)
                m -= 1

                traverseColumn(c, range(m - 1, r - 1, -1), res)
                c += 1
            else:
                traverseRow(r, range(c, n), res)
                r += 1

                traverseColumn(n - 1, range(r, m), res)
                n -= 1

            reverse = not reverse
        return res

File: leetcode/spiral-matrix.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<int> spiralOrder(vector<vector<int>>& matrix)
    {
        vector<int> res;

        int n = matrix.size(), m = matrix[0].size();
        int r = 0, c = 0;

        bool reverse = false;
        res.reserve(n * m);

        while (r < n && c < m) {
            if (reverse) {
                for (auto i = m - 1; i >= c; --i) res.push_back(matrix[n - 1][i]);
                --n;

                for (auto i = n - 1; i >= r; --i) res.push_back(matrix[i][c]);
                ++c;
            } else {
                for (auto i = c; i < m; ++i) res.push_back(matrix[r][i]);
                ++r;

                for (auto i = r; i < n; ++i) res.push_back(matrix[i][m - 1]);
                --m;
            }
            reverse = !reverse;
        }

        return res;
    }
};
